### DISCUSSION TITLE
** operartor over DictConfig shows an reportCallIssue

### DISCUSSION THREAD

**Discussion by alanlivio** [microsoft/pyright#7263] [Q&A] [answered]
OmegaConf is a popular tool for ML projects. Its main class, [DictConfig](https://github.com/omry/omegaconf/blob/5525cd5b0659311e9a2905bca3cac2fcdb5189dc/omegaconf/dictconfig.py#L57), is a collections.abc.MutableMapping.
But when ** operator over a DictConfig objs, Pylance gives a reportCallIssue with the message: `Argument expression after ** must be a mapping with an "str" key type`.

How to avoid such error? Or is it a bug on pyright?
See code bellow.

```
from omegaconf import OmegaConf as oc
from dataclasses import dataclass

@dataclass
class Config:
    param: str

ditcconf = oc.create({"param":"value"})
mycfg = Config(**ditcconf)
```

**Comment by erictraut** [comment, marked as answer]
Yes, pyright is behaving correctly here. The `DictConfig` type allows keys of type `str | bytes | int, Enum | float | bool`, but the constructor for `Config` requires a mapping with keys that are exclusively `str`.

One workaround is to write a function that converts from `DictConfig` to a `dict[str, Any]`.

```python
def make_str_dict(conf: DictConfig) -> dict[str, Any]:
    return {k: v for k, v in conf.items() if isinstance(k, str)}

mycfg = Config(**make_str_dict(dictconf))
```

Another workaround is to simply use a `typing.cast` call.

**Reply by alanlivio** [reply]
Why not using the `assert isinstance(ditcconf, dict[str, Any])` is not enoght? Example bellow.

```
from typing import Any
from omegaconf import OmegaConf as oc
from dataclasses import dataclass

@dataclass
class Config:
    param: str

ditcconf = oc.create({"param":"value"})
assert isinstance(ditcconf, dict[str, Any])
mycfg = Config(**ditcconf)
```

**Reply by erictraut** [reply]
@alanlivio, you cannot use subscripted classes in an `isinstance` call. Your code sample raises an exception at runtime.

**Reply by alanlivio** [reply]
Sorry for the confusion. After looking [at this discussion](https://github.com/python/typing/discussions/1132), I considered doing a `assert(type_inspect.get_args(ditcconf)[0],str)`, but the [get_args](https://github.com/ilevkivskyi/typing_inspect) is unfortunately empty for the [DictConfig](https://github.com/omry/omegaconf/blob/master/omegaconf/dictconfig.py).

One simplification of your solution is `ditcconf = { str(k): v for k,v in oc.create({"param":"value"}).items()}` to guarante the `dict[str, Any] `type.
